### PR TITLE
feat: add injection metadata fields to telemetry forwarder

### PR DESCRIPTION
**Pull request by sydney-tung**
# What Does This Do
This change adds additional fields to the telemetry forwarder used during injection. These fields will be used to add additional context that will be surfaced to the end user in the following Figma:

Screenshot 2025-07-03 at 10 39 35 AM

See [Injection Metadata](https://docs.google.com/document/d/1wydXj2fW0V0jaWU6W9pIbipT_AXD15tmbTtOPXFBEy4/edit?tab=t.f3vpy615c2dp) for more details.

# Motivation
Adding addition information for telemetry forwarder during injection. 

# Additional Notes

# Contributor Checklist

- Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [CODEOWNERS](https://github.com/DataDog/dd-trace-java/blob/master/.github/CODEOWNERS) file on source file addition, move, or deletion
- Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [INPLAT-614](https://datadoghq.atlassian.net/browse/INPLAT-614)

[INPLAT-614]: https://datadoghq.atlassian.net/browse/INPLAT-614?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by AlexeyKuznetsov-DD — CHANGES_REQUESTED** [review]
Looks like you accidentally committed bunch of binaries

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 1]
```diff
@@ -106,32 +130,62 @@ public void initMetaInfo(String attr, String value) {
...
+    public void setInjectResultClass(String resultClass) {
+      initMetaInfo("inject_result_class", resultClass);
+    }
+
     @Override
     public void onAbort(String reasonCode) {
       onPoint("library_entrypoint.abort", "reason:" + reasonCode);
+      setInjectResult("abort");
+      setInjectResultReason("Abort condition occurred: " + reasonCode);
+      setInjectResultClass("abort");
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: Can this (and other places) be covered with tests?

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 2]
```diff
@@ -107,15 +123,37 @@ public void initMetaInfo(String attr, String value) {
...
+
+    @Override
+    public void setInjectResultReason(String resultReason) {
+      initMetaInfo("resultReason", resultReason);
+    }
+
+    @Override
+    public void setInjectResultClass(String resultClass) {
+      initMetaInfo("resultClass", resultClass);
+    }
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: Do we need these methods at all? Since it is a thin wrapper over `initMetaInfo`?
I have a feeling it is creating more noise than value. WDYT?

File: dd-java-agent/src/main/java6/datadog/trace/bootstrap/AgentPreCheck.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 /** Special lightweight pre-main class that skips installation on incompatible JVMs. */
 public class AgentPreCheck {
+  public static final int minJavaVersion = 8;
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java6/datadog/trace/bootstrap/AgentPreCheck.java]: As constant it should be named as: `MIN_JAVA_VERSION`

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 4]
```diff
@@ -107,15 +123,37 @@ public void initMetaInfo(String attr, String value) {
...
+      initMetaInfo("result_class", resultClass);
+    }
+
     @Override
     public void onAbort(String reasonCode) {
       onPoint("library_entrypoint.abort", "reason:" + reasonCode);
       markIncomplete();
+      setInjectResult("abort");
+      setInjectResultReason(reasonCode);
+      setInjectResultClass(mapResultClass(reasonCode));
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: I would move these 3 lines into separate function and reuse it to reduce boilerplate code.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 5]
```diff
@@ -126,7 +164,27 @@ public void onFatalError(Throwable t) {
...
+
+    private String mapResultClass(String reasonCode) {
+      if (reasonCode == null) {
+        return "success";
+      }
+      switch (reasonCode) {
+        case "already_initialized":
+          return "already_instrumented";
+        case "other-java-agents":
+          return "already_instrumented";
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: Looks like these 2 `case`s can be combined like:
```
    case "already_initialized":
    case "other-java-agents":
        return "already_instrumented";
```

File: dd-java-agent/src/main/java6/datadog/trace/bootstrap/AgentPreCheck.java [hunk 6]
```diff
@@ -59,6 +60,10 @@ static void sendTelemetry(String forwarderPath, String javaVersion, String agent
             + "\"points\":[{"
             + "\"name\":\"library_entrypoint.abort\","
             + "\"tags\":[\"reason:incompatible_runtime\"]"
+            + "\"result:error\","
+            + "\"result_class:incompatible_runtime\","
+            + "\"result_reason:need to install a java version later than " + minJavaVersion + "\""
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java6/datadog/trace/bootstrap/AgentPreCheck.java]: This phrase is a bit misleading to me...
I could think that I need java > 8, but we support java 8.
Consider to use something like: `Java version 8+ required`

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java (hunk 2 above)
Comment (sydney-tung) [reply dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: I can update it to be more concise!

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 7]
```diff
@@ -53,6 +53,8 @@ public static BootstrapInitializationTelemetry createFromForwarderPath(String fo
 
   public abstract void onError(String reasonCode);
 
+  public abstract void setMetadata(String result, String resultReason, String resultClass);
+
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: I think we do not need to have it on base class, I feel that we can declare it private in `JsonBased` class only.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 8]
```diff
@@ -126,7 +135,31 @@ public void onFatalError(Throwable t) {
...
       onPoint("library_entrypoint.error", "error_type:" + reasonCode);
+      setMetadata("error", reasonCode, mapResultClass(reasonCode));
+    }
+
+    @Override
+    public void setMetadata(String result, String resultReason, String resultClass) {
+      initMetaInfo("result", result);
+      initMetaInfo("result_reason", resultReason);
+      initMetaInfo("result_class", resultClass);
+    }
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: I have my personal opinion, that it would be better to rearrange this code a bit...
You would need to change related places.
  ```suggestion
    @Override
    public void setMetadata(String result, String resultClass, String resultReason) {
      initMetaInfo("result", result);
      initMetaInfo("result_class", resultClass);
      initMetaInfo("result_reason", resultReason);
    }
```

**Review by AlexeyKuznetsov-DD — CHANGES_REQUESTED** [review]

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 9]
```diff
@@ -38,7 +38,66 @@ class BootstrapInitializationTelemetryTest extends Specification {
...
+    when:
+    initTelemetry.initMetaInfo("runtime_name", "java")
+    initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
+
+    initTelemetry.onFatalError(new Exception("foo"))
+    initTelemetry.finish()
+
+    then:
+    capture.json() == '{"metadata":{"runtime_name":"java","runtime_version":"1.8.0_382","result":"error","result_class":"internal_error","result_reason":"foo"},"points":[{"name":"library_entrypoint.error","tags":["error_type:java.lang.Exception"]}]}'
+  }
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: Looks like this test can be combined with existing test `def "incomplete on fatal error"()`?

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 10]
```diff
@@ -38,7 +38,66 @@ class BootstrapInitializationTelemetryTest extends Specification {
...
+
+  def "test abort"() {
+    when:
+    initTelemetry.initMetaInfo("runtime_name", "java")
+    initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: I have a feeling that this 2 lines can be moved to `setup()` method (it is the same as `@BeforeEach` in JUnit).
And reused for all tests.

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 11]
```diff
@@ -38,7 +38,66 @@ class BootstrapInitializationTelemetryTest extends Specification {
...
+    initTelemetry.finish()
+
+    then:
+    capture.json() == '{"metadata":{"runtime_name":"java","runtime_version":"1.8.0_382","result":"abort","result_class":"unsupported_binary","result_reason":"jdk_tool"},"points":[{"name":"library_entrypoint.abort","tags":["reason:jdk_tool"]}]}'
+  }
+
+  def "test success"() {
+    when:
+    initTelemetry.initMetaInfo("runtime_name", "java")
+    initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: I would move this test to be the first one, it is optional, but usually developers organize tests like:
* happy path
* corner case 1
   ...
* corner case N

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 12]
```diff
@@ -38,7 +38,66 @@ class BootstrapInitializationTelemetryTest extends Specification {
...
+
+    initTelemetry.finish()
+
+    then:
+    capture.json() == '{"metadata":{"runtime_name":"java","runtime_version":"1.8.0_382","result":"success","result_class":"success","result_reason":"Successfully configured ddtrace package"},"points":[{"name":"library_entrypoint.complete"}]}'
+  }
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: I think we can introduce patter or small private function to get this JSON and use in tests. This will reduce boilerplate.

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 13]
```diff
@@ -38,7 +38,66 @@ class BootstrapInitializationTelemetryTest extends Specification {
...
+    then:
+    capture.json() == '{"metadata":{"runtime_name":"java","runtime_version":"1.8.0_382","result":"abort","result_class":"already_instrumented","result_reason":"other-java-agents"},"points":[{"name":"library_entrypoint.abort","tags":["reason:other-java-agents"]}]}'
+  }
+
+  def "test abort unknown"() {
+    when:
+    initTelemetry.initMetaInfo("runtime_name", "java")
+    initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
+
+    initTelemetry.onAbort("foo")
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: You can combine these 2 tests into one using `where` feature of Spock framework.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 14]
```diff
@@ -126,7 +130,30 @@ public void onFatalError(Throwable t) {
 
     @Override
     public void onError(String reasonCode) {
+      error = true;
       onPoint("library_entrypoint.error", "error_type:" + reasonCode);
+      setMetadata("error", mapResultClass(reasonCode), reasonCode);
+    }
+
+    public void setMetadata(String result, String resultClass, String resultReason) {
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: Do we need this method to be `public`? If not, let's make it `private`.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 15]
```diff
@@ -126,7 +130,30 @@ public void onFatalError(Throwable t) {
...
+      initMetaInfo("result", result);
+      initMetaInfo("result_class", resultClass);
+      initMetaInfo("result_reason", resultReason);
+    }
+
+    private String mapResultClass(String reasonCode) {
+      if (reasonCode == null) {
+        return "success";
+      }
+      switch (reasonCode) {
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java]: Minor: I would suggest to add an empty line for readability.
```suggestion
      if (reasonCode == null) {
        return "success";
      }

      switch (reasonCode) {
```

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 16]
```diff
@@ -143,6 +170,9 @@ public void markIncomplete() {
 
     @Override
     public void finish() {
+      if (!this.incomplete && !this.error) {
+        setMetadata("success", "success", "Successfully configured ddtrace package");
+      }
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java:177]: Minor: Add empty line for readability:
```suggestion
      if (!this.incomplete && !this.error) {
        setMetadata("success", "success", "Successfully configured ddtrace package");
      }

```

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 17]
```diff
@@ -27,18 +27,47 @@ class BootstrapInitializationTelemetryTest extends Specification {
 
     this.initTelemetry = initTelemetryProxy
     this.capture = capture
+
+    this.initTelemetry.initMetaInfo("runtime_name", "java")
+    this.initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy]: I would group `initTelemetry` together. Just move these 2 lines to be after line 28

**Review by AlexeyKuznetsov-DD — CHANGES_REQUESTED** [review]
LGTM, but minor cleanup required.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 18]
```diff
@@ -126,7 +130,31 @@ public void onFatalError(Throwable t) {
...
+
+    private String mapResultClass(String reasonCode) {
+      if (reasonCode == null) {
+        return "success";
+      }
+
+      switch (reasonCode) {
+        case "already_initialized":
+        case "other-java-agents":
+          return "already_instrumented";
```
Comment (dougqh) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java:151]: "other-java-agents" isn't the same as already instrumented.

"already_initialized" indicates that the dd-java-agent has already installed through some other means.

"other-java-agents" indicates that dd-java-agent wasn't installed because there is a potentially conflicting agent present.

File: .gitignore [hunk 19]
```diff
@@ -60,6 +60,7 @@ replay_pid*
 .bsp/
 **/src/*/generated/*
 !dd-java-agent/benchmark/releases/*.jar
+*.jar
```
Comment (PerfectSlayer) [inline .gitignore:63]: We do want jar, like maven and gradle wrappers. Can you revert this part please?

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java [hunk 20]
```diff
@@ -126,7 +130,32 @@ public void onFatalError(Throwable t) {
 
     @Override
     public void onError(String reasonCode) {
+      error = true;
       onPoint("library_entrypoint.error", "error_type:" + reasonCode);
+      setMetadata("error", mapResultClass(reasonCode), reasonCode);
+    }
+
+    private void setMetadata(String result, String resultClass, String resultReason) {
```
Comment (PerfectSlayer) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/BootstrapInitializationTelemetry.java:138]: I would rename the result helper method and use overloading for clarity. Something like:

```suggestion
    private void setMetaResult(String result, String resultCode) {
      setMetaResult(result,  mapResultClass(reasonCode), reasonCode);
    }
    
    private void setMetaResult(String result, String resultClass, String resultReason) {
    
    }
```

WDYT?

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 21]
```diff
@@ -26,19 +26,47 @@ class BootstrapInitializationTelemetryTest extends Specification {
     initTelemetryProxy.setAdaptee(initTelemetry)
 
     this.initTelemetry = initTelemetryProxy
+    this.initTelemetry.initMetaInfo("runtime_name", "java")
+    this.initTelemetry.initMetaInfo("runtime_version", "1.8.0_382")
```
Comment (PerfectSlayer) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy:30]: If this version is coming from you current setup or the current CI JDK version, you can rely on it for testing

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy [hunk 22]
```diff
@@ -76,6 +106,10 @@ class BootstrapInitializationTelemetryTest extends Specification {
     !capture.json().contains("library_entrypoint.complete")
   }
 
+  private String json(String result, String resultClass, String resultReason, List points) {
+    return """{"metadata":{"runtime_name":"java","runtime_version":"1.8.0_382","result":"${result}","result_class":"${resultClass}","result_reason":"${resultReason}"},"points":${new groovy.json.JsonBuilder(points)}}"""
```
Comment (PerfectSlayer) [inline dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy:110]: Same here

**Review by PerfectSlayer — CHANGES_REQUESTED** [review]
Left a bunch of comments.
The request for change is only for the `.gitignore`  🙏

File: dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy (hunk 21 above)
Comment (AlexeyKuznetsov-DD) [reply dd-java-agent/src/test/groovy/datadog/trace/bootstrap/BootstrapInitializationTelemetryTest.groovy:30]: This is 100% synthetic test, no need in real data.
